### PR TITLE
#25 - Workflow visualization for "WorkflowEditor"

### PR DESCRIPTION
**Pull request by kvolkovich-sc**
### Release notes

```release-notes
Added workflow visualization using "graphviz" (https://www.graphviz.org/)
```

### CODE REVIEW

File: crud-editor/src/components/WorkflowGraph/WorkflowGraph.react.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    src += `\trankdir=LR;\n`;
+    src += `\tedge [fontname="Helvetica"];\n`;
+    // eslint-disable-next-line max-len
+    src += `\tnode [shape = rect fillcolor="#b71c1c" margin="0.2,0.1" color="transparent" fontname="Helvetica" style="rounded,filled"];\n`;
+    src += `\t${finalStates.map(state => `"${state}"`).join('')}\n`;
+    src += `\tnode [fillcolor="#14892c"];\n`;
+    src += `\t"${initialState}"\n`;
+    src += `\tnode [fillcolor="#0277bd"];\n`;
+    src += transitions.
+      filter(({ from, to, event }) => (from !== null && to !== null && event)).
```
Comment (estambakio-sc) [inline crud-editor/src/components/WorkflowGraph/WorkflowGraph.react.js]: Invalid transition also includes empty strings for `from`, `to` and `null`. Better condition is `from && to && event`.

File: crud-editor/src/components/WorkflowGraph/WorkflowGraph.react.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    // DOT language used by graphviz: https://graphviz.gitlab.io/_pages/doc/info/lang.html
+    let { transitions, initialState, finalStates } = schema;
+
+    let src = '';
+    src += `digraph finite_state_machine {\n`;
+    src += `\trankdir=LR;\n`;
+    src += `\tedge [fontname="Helvetica"];\n`;
+    // eslint-disable-next-line max-len
+    src += `\tnode [shape = rect fillcolor="#b71c1c" margin="0.2,0.1" color="transparent" fontname="Helvetica" style="rounded,filled"];\n`;
+    src += `\t${finalStates.map(state => `"${state}"`).join('')}\n`;
```
Comment (estambakio-sc) [inline crud-editor/src/components/WorkflowGraph/WorkflowGraph.react.js]: `join('')` shouldn't it be space?

File: crud-editor/package.json [hunk 3]
```diff
@@ -1,57 +1,59 @@
...
+    "build": "cross-env NODE_ENV=development showroom-scan ./src && webpack --config config/webpack.config.showroom.js",
+    "lint-fix": "node ./node_modules/eslint/bin/eslint.js ./src --config ./config/.eslintrc --fix"
+  },
+  "dependencies": {
+    "@opuscapita/react-select": "1.0.0-oc.0",
+    "codemirror": "^5.33.0",
+    "kvolkovich-sc-react-codemirror": "^1.0.1",
+    "lodash": "^4.17.4",
+    "react-bootstrap": "0.32.0",
+    "react-codemirror2": "^3.0.7",
```
Comment (estambakio-sc) [inline crud-editor/package.json]: Please remove `react-codemirror2` and fix packages versions.

**Comment by kvolkovich-sc** [comment]
Done appropriate changes according to @estambakio-sc code review.
@nstoliar-sc, please test. [Demo](https://opuscapita.github.io/fsm-workflow/branches/issue-25(slash)workflow-visualization/crud-editor/?currentComponentName=WorkflowEditor&maxContainerWidth=100%25)

**Review by estambakio-sc — APPROVED** [review]

**Comment by estambakio-sc** [comment]
New [demo link](https://opuscapita.github.io/fsm-workflow/branches/issue-23(slash)workflow-editor/crud-editor/?currentComponentName=WorkflowEditor)

**Pull request merged**